draft-ietf-mls-protocol.md

```diff
@@ -1698,85 +1684,110 @@ struct {
-
-        case commit:
-            opaque parent_hash<V>;
-    }
-
     Extension extensions<V>;
 
-    select (leaf_node_source) {
+    ClientInfoSource client_info_source;
+    select (client_info_source) {
         case key_package:
-            struct{};
-
+        case add:
+            Lifetime lifetime;
         case update:
             opaque group_id<V>;
-
         case commit:
+            opaque parent_hash<V>;
             opaque group_id<V>;
     }
-} LeafNodeTBS;
+} ClientInfoTBS;
+
+LeafNode ClientInfo;
 ~~~~~
 
-The `public_key` field conains an HPKE public key whose private key is held only
-by the member occupying this leaf.  The `credential` contains authentication
-information for this member, as described in {{credentials}}.
+The `public_key` field conains an ephemeral HPKE public key whose private key
+is held only by the issuer of the KeyPackage (when used inside a KeyPackage),
+or the member occupying the referenced leaf (when used as a LeafNode).
+The `credential` contains permanent or semi-permanent authentication
+information for this member or potential member, as described in {{credentials}}.
 
 The `capabilities` field indicates what protocol versions, ciphersuites,
 protocol extensions, and non-default proposal types are supported by a client.
 Proposal types defined in this document are considered "default" and thus need
-not be listed.  Extensions that appear in the `extensions` field of a LeafNode
+not be listed.  Extensions that appear in the `extensions` field of a ClientInfo
 MUST be included in the `extensions` field of the `capabilities` field.
 
-The `leaf_node_source` field indicates how this LeafNode came to be added to the
-tree.
+The `client_info_source` field indicates the context in which the ClientInfo
+object is used.
 
-In the case where the leaf was added to the tree based on a pre-published
-KeyPackage, the `lifetime` field represents the times between which clients will
-consider a LeafNode valid.  These times are represented as absolute times,
-measured in seconds since the Unix epoch (1970-01-01T00:00:00Z).  Applications
-MUST define a maximum total lifetime that is acceptable for a LeafNode, and
-reject any LeafNode where the total lifetime is longer than this duration.
+In a KeyPackage or an Add proposal, the `lifetime` field represents the times
+between which clients will consider the KeyPackage or LeafNode (public_key and credential) valid.  These
+times are represented as absolute times, measured in seconds since the
+Unix epoch (1970-01-01T00:00:00Z).  Applications MUST define a maximum total
+lifetime that is acceptable for a ClientInfo object, and reject any ClientInfo
+object where the total lifetime is longer than this duration.
 
-In the case where the leaf node was inserted into the tree via a Commit message,
-the `parent_hash` field contains the parent hash for this leaf node (see
-{{parent-hash}}).
+In the case where the ClientInfo refers to a leaf node inserted into the tree
```

Comment:
```suggestion
In the case where the ClientInfo represents a leaf node inserted into the tree
```

Reply:
Yours is better. What do you think of:
...represents a leaf node once it is inserted into the tree.
?